pip/req/req_cache.py

```diff
@@ -0,0 +1,217 @@
+from __future__ import absolute_import
+
+from collections import defaultdict
+import logging
+import os.path
+import shutil
+import tempfile
+
+from pip.download import url_to_path
+from pip.utils import display_path, ensure_dir, rmtree, _make_build_dir
+
+
+logger = logging.getLogger(__name__)
+
+
+class RequirementCache(object):
+    """A cache of requirements.
+
+    :attr delete: May be set to False to disable deleting the cache.
+
+    Design
+    ------
+
+    The cache is responsible for holding requirements - be those local paths,
+    tarballs, VCS repositories or package names. The constraints placed on
+    selection of requirements are handled by the RequirementSet, and must not
+    be pushed down into the cache - because the cache will return the same
+    requirement each time it is referenced. InstallRequirement is being
+    broken into the cachable component (CachedRequirement) and the uncachable
+    per-evaluation component (InstallRequirement).
+
+    RequirementCache owns CachedRequirements, so the primary interface is
+    to ask the cache to perform an operation when cache state is needed.
+    """
+
+    def __init__(self, path=None, delete=None, src_dir=None):
+        """Create a RequirementCache.
+
+        :param path: The path to store working data - wheels, sdists, etc. If
+            None then a temp dir will be made on __enter__.
+        :param delete: If False then path will not be deleted on __exit__.
+            If None, then path will be deleted on __exit__ only if path was
+            None. If True then path will be deleted on __exit__.
+        :param src_dir: The parent path that editable non-local requirements
+            will be checked out under. Will be created on demand if needed.
+        """
+        # If we were not given an explicit directory, and we were not given an
+        # explicit delete option, then we'll default to deleting.
+        if path is None and delete is None:
+            delete = True
+        self._path = path
+
+        self.path = None
+        self.delete = delete
+        self.src_dir = os.path.abspath(src_dir) if src_dir else None
+        self._urls = None
+
+    def __repr__(self):
+        return "<{} {!r}>".format(self.__class__.__name__, self.path)
+
+    def __enter__(self):
+        if self._path is None:
+            # We realpath here because some systems have their default tmpdir
+            # symlinked to another directory.  This tends to confuse build
+            # scripts, so we canonicalize the path by traversing potential
+            # symlinks here.
+            self.path = os.path.realpath(tempfile.mkdtemp(prefix="pip-build-"))
+            # If we were not given an explicit directory, and we were not given
+            # an explicit delete option, then we'll default to deleting.
+            if self.delete is None:
+                self.delete = True
```

Comment:
Unless there's something I'm missing that means self._path could be unset in the meantime, this seems to duplicate lines 49-50 and 54